### ISSUE TITLE
Create dependencies and later delete all applications causes an internal server error

### ISSUE THREAD

**Issue by carlosthe19916** [konveyor/tackle-application-inventory#40]
**Steps to reproduce the bug**
- Create applications using `POST /application`
- Create dependencies using `POST /applications-dependency`
- Delete all applications
- Create a new application and try to execute`GET /applications-dependency` you will see an error.

I'm attaching a video to better explain the steps to be done:

https://user-images.githubusercontent.com/2582866/115597332-71749200-a2d9-11eb-8473-e5c6c0688c0c.mp4

This is the log I see in the backend:

```
application-inventory_1     | 2021-04-21 17:38:39,030 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-2) HTTP Request to /application-inventory/applications-dependency?page=0&size=1000 failed, error id: b0bd9977-a7dc-4af9-94e9-96a5996530b1-1: org.jboss.resteasy.spi.UnhandledException: javax.persistence.EntityNotFoundException: Unable to find io.tackle.applicationinventory.entities.Application with id 1
application-inventory_1     |   at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:106)
application-inventory_1     |   at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:372)
application-inventory_1     |   at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:218)
application-inventory_1     |   at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:519)
application-inventory_1     |   at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
application-inventory_1     |   at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
application-inventory_1     |   at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
application-inventory_1     |   at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
application-inventory_1     |   at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
application-inventory_1     |   at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
application-inventory_1     |   at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:138)
application-inventory_1     |   at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.access$000(VertxRequestHandler.java:41)
application-inventory_1     |   at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:93)
application-inventory_1     |   at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2415)
application-inventory_1     |   at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
application-inventory_1     |   at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
application-inventory_1     |   at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
application-inventory_1     |   at java.lang.Thread.run(Thread.java:834)
application-inventory_1     |   at org.jboss.threads.JBossThread.run(JBossThread.java:501)
application-inventory_1     |   at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:519)
application-inventory_1     |   at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
application-inventory_1     | Caused by: javax.persistence.EntityNotFoundException: Unable to find io.tackle.applicationinventory.entities.Application with id 1
application-inventory_1     |   at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder$JpaEntityNotFoundDelegate.handleEntityNotFound(FastBootEntityManagerFactoryBuilder.java:193)
application-inventory_1     |   at org.hibernate.event.internal.DefaultLoadEventListener.load(DefaultLoadEventListener.java:216)
application-inventory_1     |   at org.hibernate.event.internal.DefaultLoadEventListener.proxyOrLoad(DefaultLoadEventListener.java:332)
application-inventory_1     |   at org.hibernate.event.internal.DefaultLoadEventListener.doOnLoad(DefaultLoadEventListener.java:108)
application-inventory_1     |   at org.hibernate.event.internal.DefaultLoadEventListener.onLoad(DefaultLoadEventListener.java:74)
application-inventory_1     |   at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:104)
application-inventory_1     |   at org.hibernate.internal.SessionImpl.fireLoadNoChecks(SessionImpl.java:1186)
application-inventory_1     |   at org.hibernate.internal.SessionImpl.internalLoad(SessionImpl.java:1051)
application-inventory_1     |   at org.hibernate.type.EntityType.resolveIdentifier(EntityType.java:697)
application-inventory_1     |   at org.hibernate.type.EntityType.resolve(EntityType.java:464)
application-inventory_1     |   at org.hibernate.type.ManyToOneType.resolve(ManyToOneType.java:240)
application-inventory_1     |   at org.hibernate.engine.internal.TwoPhaseLoad$EntityResolver.lambda$static$0(TwoPhaseLoad.java:605)
application-inventory_1     |   at org.hibernate.engine.internal.TwoPhaseLoad.initializeEntityEntryLoadedState(TwoPhaseLoad.java:248)
application-inventory_1     |   at org.hibernate.engine.internal.TwoPhaseLoad.initializeEntity(TwoPhaseLoad.java:182)
application-inventory_1     |   at org.hibernate.engine.internal.TwoPhaseLoad.initializeEntity(TwoPhaseLoad.java:151)
application-inventory_1     |   at org.hibernate.loader.Loader.initializeEntitiesAndCollections(Loader.java:1199)
application-inventory_1     |   at org.hibernate.loader.Loader.processResultSet(Loader.java:1001)
application-inventory_1     |   at org.hibernate.loader.Loader.doQuery(Loader.java:959)
application-inventory_1     |   at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:349)
application-inventory_1     |   at org.hibernate.loader.Loader.doList(Loader.java:2849)
application-inventory_1     |   at org.hibernate.loader.Loader.doList(Loader.java:2831)
application-inventory_1     |   at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2663)
application-inventory_1     |   at org.hibernate.loader.Loader.list(Loader.java:2658)
application-inventory_1     |   at org.hibernate.loader.hql.QueryLoader.list(QueryLoader.java:506)
application-inventory_1     |   at org.hibernate.hql.internal.ast.QueryTranslatorImpl.list(QueryTranslatorImpl.java:400)
application-inventory_1     |   at org.hibernate.engine.query.spi.HQLQueryPlan.performList(HQLQueryPlan.java:219)
application-inventory_1     |   at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1414)
application-inventory_1     |   at org.hibernate.query.internal.AbstractProducedQuery.doList(AbstractProducedQuery.java:1625)
application-inventory_1     |   at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1593)
application-inventory_1     |   at org.hibernate.query.Query.getResultList(Query.java:165)
application-inventory_1     |   at io.quarkus.hibernate.orm.panache.common.runtime.CommonPanacheQueryImpl.list(CommonPanacheQueryImpl.java:239)
application-inventory_1     |   at io.quarkus.hibernate.orm.panache.runtime.PanacheQueryImpl.list(PanacheQueryImpl.java:149)
application-inventory_1     |   at io.tackle.commons.resources.ListFilteredResource.list(ListFilteredResource.java:226)
application-inventory_1     |   at io.tackle.commons.resources.ListFilteredResource.list(ListFilteredResource.java:194)
application-inventory_1     |   at io.tackle.applicationinventory.resources.ApplicationsDependencyListFilteredResource.listHal(ApplicationsDependencyListFilteredResource.java:46)
application-inventory_1     |   at java.lang.reflect.Method.invoke(Method.java:566)
application-inventory_1     |   at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
application-inventory_1     |   at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
application-inventory_1     |   at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:643)
application-inventory_1     |   at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:507)
application-inventory_1     |   at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:457)
application-inventory_1     |   at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
application-inventory_1     |   at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:459)
application-inventory_1     |   at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:419)
application-inventory_1     |   at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:393)
application-inventory_1     |   at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:68)
application-inventory_1     |   at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
application-inventory_1     |   ... 17 more
application-inventory_1     | 
```

**Comment by mrizzi** [comment]
Well spotted, thanks @carlosthe19916.

Checking your video, I think it's worth explaining that calling `/application-inventory/applications-dependency?page=0&size=1000` when opening the "Manage dependencies" is not the recommended approach because this would load **all** the dependencies for all the applications and not just the one related to the application the user wants to manage the dependencies for.

Referring to the usage guide for the [`/applications-dependency` endpoint](https://github.com/konveyor/tackle-application-inventory/blob/main/USAGE.md#applications-dependency-endpoint), the loading of the dependencies for that modal is better be divided into two more granular calls:

1. retrieve only the northbound dependencies for the application following the [Retrieve northbound dependencies for an application](https://github.com/konveyor/tackle-application-inventory/blob/main/USAGE.md#retrieve-northbound-dependencies-for-an-application) guide
1. retrieve only the southbound dependencies for the application following the [Retrieve southbound dependencies for an application](https://github.com/konveyor/tackle-application-inventory/blob/main/USAGE.md#retrieve-southbound-dependencies-for-an-application) guide

**Comment by carlosthe19916** [comment]
@mrizzi Thanks for the information about how to fetch dependencies only for one application using the queryParam `from.id` or `to.id`. I did it that way because I didn't know it was possible to fetch dependencies for only one application; I must admit I didn't read the `USAGE.md` file.

In my opinion Markdown files are not the best way of documenting the endpoints. For instance, I used the openapi.json definition of the microservice to know which is the current status of `tackle-application-inventory` and this is what I saw:

![Screenshot from 2021-04-22 11-14-42](https://user-images.githubusercontent.com/2582866/115689922-d7562d80-a35c-11eb-9354-03aa2d21b9e8.png)

The openapi.json doesn't contain the queryParams you mentioned. I do think it is worth inventing time on it and setting it on the roadmap.
